### PR TITLE
HIV-388 : Integrating surge report into frontend

### PR DESCRIPTION
**Pull request by donaldkibet**
Integrating Surge Report into front end

### CODE REVIEW

File: src/app/data-analytics-dashboard/hiv/surge/surge-report.component.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    this.loadParametersFromUrl();
+  }
+
+  public generateReport() {
+    this.storeParamsInUrl();
+    super.generateReport();
+  }
+
+  public storeParamsInUrl() {
+    this.getLocationsSelected();
```
Comment (nkimaina) [inline src/app/data-analytics-dashboard/hiv/surge/surge-report.component.ts]: rename to 'getSelectedLocations'

File: src/app/data-analytics-dashboard/hiv/surge/surge-report.component.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  public loadParametersFromUrl() {
+    const path = this.router.parseUrl(this.location.path());
+
+    if (path.queryParams['state']) {
+      const state = rison.decode(path.queryParams['state']);
+      this.yearWeek = state.year_week;
+    }
+
+    if (path.queryParams['state']) {
+      this.generateReport();
```
Comment (nkimaina) [inline src/app/data-analytics-dashboard/hiv/surge/surge-report.component.ts]: move this call to line 62 as there's an if statement checking for presence of state object

File: src/app/etl-api/etl-api.module.ts [hunk 3]
```diff
@@ -78,7 +82,9 @@ import { OncologySummaryResourceService } from './oncology-summary-resource.serv
     HivEnhancedReportService,
     HivEnhancedReportService,
     OncologySummaryResourceService,
-    RadiologyImagingResourceService
```
Comment (nkimaina) [inline src/app/etl-api/etl-api.module.ts:81]: why are we removing thise radiology resource service? some code might break somewhere

File: src/app/etl-api/surge-resource.service.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+          };
+          return Observable.of(errorObj);
+        }),
+        map((response: Response) => {
+          return response;
+        })
+      );
+  }
+
+  public getSurgePatientList(params: any): Observable<any> {
```
Comment (nkimaina) [inline src/app/etl-api/surge-resource.service.ts]: rename to getSurgeWeeklyPatientList or if you plan to re-use it to get the daily report, then rename line 21 appropriately

File: src/app/hiv-care-lib/surge-report/surge-report-base.component.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+export class SurgeReportBaseComponent implements OnInit {
+  public params: any;
+  public indicators: string;
+  public selectedIndicators = [];
+  public surgeWeeklyReportSummaryData: any = [];
+  public columnDefs: any = [];
+  public enabledControls = 'weekControl';
+  public reportName = 'Surge Weekly Report';
+  public currentView = 'weekly';
+  public isReleased = false;
```
Comment (nkimaina) [inline src/app/hiv-care-lib/surge-report/surge-report-base.component.ts]: We don't need this variable, as it was used for MOH report where we have unreleased and released reports

File: src/app/hiv-care-lib/surge-report/surge-report-base.component.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        this.displayTabluarFilters = data.displayTabluarFilters;
+      }
+    );
+   }
+
+   ngOnInit() {
+   }
+
+  public getSurgeWeeklyReport(params: any) {
+    this.surgeReport.getSurgeWeeklyReport(params).subscribe(data => {
```
Comment (nkimaina) [inline src/app/hiv-care-lib/surge-report/surge-report-base.component.ts:84]: setting of isLoading to 'true', should actually be done between line 83 and 84

File: src/app/hiv-care-lib/surge-report/surge-report-patientlist.component.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+      });
+  }
+
+  private getPatientList(params: any) {
+    this.surgeResource.getSurgePatientList(params)
+        .subscribe(
+          (data) => {
+            this.isLoading = false;
+            this.patientData = data.results.results;
+            this.isLoading = false;
```
Comment (nkimaina) [inline src/app/hiv-care-lib/surge-report/surge-report-patientlist.component.ts]: this is already set to false in line 48

File: src/app/etl-api/etl-api.module.ts (hunk 3 above)
Comment (donaldkibet) [reply src/app/etl-api/etl-api.module.ts:81]: reinstated this line back

File: src/app/hiv-care-lib/surge-report/surge-report-base.component.ts (hunk 5 above)
Comment (donaldkibet) [reply src/app/hiv-care-lib/surge-report/surge-report-base.component.ts]: used this variable to hide the daily surge report controls

File: src/app/clinic-dashboard/hiv/hiv-program.routes.ts [hunk 8]
```diff
@@ -183,6 +186,14 @@ const routes: Routes = [
...
         path: 'department-select',
         component : ChangeDepartmentComponent
+    },
+    {
+        path: 'surge-reports',
+        component: SurgeReportComponent
+    },
+    {
+        path: 'surge-reports/surge-report-patientlist',
+        component: SurgeReportPatientlistComponent
```
Comment (fatmali) [inline src/app/clinic-dashboard/hiv/hiv-program.routes.ts]: Rename this to SurgeReportPatientListComponent. (camel-case)

File: src/app/data-analytics-dashboard/hiv/data-analytics-hiv.module.ts [hunk 9]
```diff
@@ -68,7 +69,8 @@ import {
     Moh731ReportComponent,
     Moh731MonthlyVizComponent,
     HivCareComparativeAnalyticsComponent,
-    HivSummaryMonthlyIndicatorsComponent
+    HivSummaryMonthlyIndicatorsComponent,
+    SurgeReportComponent
```
Comment (fatmali) [inline src/app/data-analytics-dashboard/hiv/data-analytics-hiv.module.ts:75]: This component has been also declared in hiv-program.module.ts. You should declare shared components in a shared module to be shared across the app.

File: src/app/data-analytics-dashboard/hiv/surge/surge-report.component.ts [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+    for (let i = 0; i < locationUuids.length; i++) {
+      if (i === 0) {
+        selectedLocations = selectedLocations + (locationUuids[0] as any).value;
+      } else {
+        selectedLocations = selectedLocations + ',' + (locationUuids[i] as any).value;
+      }
+    }
+    return selectedLocations;
+  }
```
Comment (fatmali) [inline src/app/data-analytics-dashboard/hiv/surge/surge-report.component.ts:105]: have a look at Array.join() and Array.map(), here you can use `selectedLocations = locationUuids.map(location => location.value).join(",")`

File: src/app/hiv-care-lib/surge-report/surge-report-patientlist.component.ts [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+      latest_vl_date: 'Latest VL Date',
+      previous_vl: 'Previous VL',
+      previous_vl_date: 'Prevoius VL Date',
+      // baseline: 'Baseline status as at 11th may 2019',
+      cur_status: 'Current Status',
+      active_to_ltfu_count: 'Active to LTFU Count',
+      // death_date: 'Death date',
+      // intervention_name: 'Intervention Name',
+      // intervention_date: 'Intervention Date',
+      // education: 'Level of Education',
```
Comment (fatmali) [inline src/app/hiv-care-lib/surge-report/surge-report-patientlist.component.ts]: Remove this columns if they are not being used.

**Review by fatmali — CHANGES_REQUESTED** [review]

File: src/app/data-analytics-dashboard/hiv/surge/surge-report.component.ts (hunk 1 above)
Comment (donaldkibet) [reply src/app/data-analytics-dashboard/hiv/surge/surge-report.component.ts]: renamed it

File: src/app/hiv-care-lib/surge-report/surge-report-patientlist.component.ts (hunk 7 above)
Comment (donaldkibet) [reply src/app/hiv-care-lib/surge-report/surge-report-patientlist.component.ts]: made the changes

File: src/app/clinic-dashboard/hiv/surge-report/surge-report.component.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    public surgeReport: SurgeResourceService) {
+    super(router, route, surgeReport);
+  }
+
+  ngOnInit() {
+    this.route.queryParams.subscribe(
+      (params: any) => {
+        switch (params.currentView) {
+          case 'daily':
+            if (params) {
```
Comment (nkimaina) [inline src/app/clinic-dashboard/hiv/surge-report/surge-report.component.ts]: it's more common for people to combine the 2 null checks under one if statement i.e if(params && params._date) { ... }

File: src/app/clinic-dashboard/hiv/surge-report/surge-report.component.ts [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+              if (params._date) {
+                this.isLoading = true;
+                this.params = params;
+                this.getSurgeDailyReport(params);
+              }
+            }
+            break;
+          case 'weekly':
+            if (params) {
+              if (params.year_week) {
```
Comment (nkimaina) [inline src/app/clinic-dashboard/hiv/surge-report/surge-report.component.ts]: do the same as the one for dates

File: src/app/data-analytics-dashboard/hiv/data-analytics-hiv.routes.ts [hunk 14]
```diff
@@ -22,6 +22,8 @@ import { Moh731MonthlyVizComponent } from './moh-731-monthly-viz/moh-731-monthly
 import {
   DashboardsViewerComponent
 } from '../../kibana-lib';
+import { SurgeReportComponent } from './surge/surge-report.component';
+import { SurgeReportPatientListComponent } from 'src/app/hiv-care-lib/surge-report/surge-report-patientlist.component';
```
Comment (drizzentic) [inline src/app/data-analytics-dashboard/hiv/data-analytics-hiv.routes.ts]:
```suggestion
import { SurgeReportPatientListComponent } from 'src/app/hiv-care-lib/surge-report/surge-report-patient-list.component';
```

File: src/app/hiv-care-lib/hiv-care-lib.module.ts [hunk 15]
```diff
@@ -87,6 +87,9 @@ import {
   ProgramWorkFlowResourceService
 } from '../openmrs-api/program-workflow-resource.service';
 import { HighchartsStatic } from 'angular2-highcharts/dist/HighchartsService';
+import { SurgeReportTabularComponent } from './surge-report/surge-report-tabular.component';
+import { SurgeReportBaseComponent } from './surge-report/surge-report-base.component';
+import { SurgeReportPatientListComponent } from './surge-report/surge-report-patientlist.component';
```
Comment (drizzentic) [inline src/app/hiv-care-lib/hiv-care-lib.module.ts]:
```suggestion
import { SurgeReportPatientListComponent } from './surge-report/surge-report-patient-list.component';
```

File: src/app/hiv-care-lib/surge-report/surge-report-base.component.html [hunk 16]
```diff
@@ -0,0 +1,48 @@
+<h4 class="component-title text-success"><span class="fa fa-area-chart"></span> Surge Reports</h4>
```
Comment (drizzentic) [inline src/app/hiv-care-lib/surge-report/surge-report-base.component.html]:
```suggestion
<h4 class="component-title text-success"><span class="fa fa-area-chart"></span>Surge Reports</h4>
```

File: src/app/hiv-care-lib/surge-report/surge-report-base.component.html [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+<div class="loader" *ngIf="isLoading">
+    <span><i class="fa fa-spinner fa-spin"> </i>Loading...</span>
```
Comment (drizzentic) [inline src/app/hiv-care-lib/surge-report/surge-report-base.component.html]:
```suggestion
    <span><i class="fa fa-spinner fa-spin"></i>Loading...</span>
```

File: src/app/hiv-care-lib/surge-report/surge-report-base.component.html [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+    </h4>
+    <p>
+        <small>{{errorMessage}}</small>
+    </p>
+</div>
+
+
+<div>
+    <p-tabView>
+        <p-tabPanel header="Report View" leftIcon="fa-file-pdf-o " [selected]="currentViewBelow === 'pdf'">
```
Comment (drizzentic) [inline src/app/hiv-care-lib/surge-report/surge-report-base.component.html]:
```suggestion
        <p-tabPanel header="Report View" leftIcon="fa-file-pdf-o" [selected]="currentViewBelow === 'pdf'">
```

File: src/app/hiv-care-lib/surge-report/surge-report-base.component.spec.ts [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+  });
+
+  it('should be injected', () => {
+    fixture.detectChanges();
+    expect(fixture.componentInstance).toBeTruthy();
+    expect(fixture.componentInstance.surgeReport instanceof SurgeResourceService).toBeTruthy();
+  });
+
+  it('should generate the surge weekly report when given the correct parameters', () => {
+
```
Comment (drizzentic) [inline src/app/hiv-care-lib/surge-report/surge-report-base.component.spec.ts]: What's the purpose of this test? Remove it or refactor if possible

File: src/app/hiv-care-lib/surge-report/surge-report-patientlist.component.html [hunk 20]
```diff
@@ -0,0 +1,24 @@
+<div>
+  <button class="btn btn-primary" (click)="goBack()">
+    <span class="glyphicon glyphicon-arrow-left"></span>
+    <span>Go back to tabular view</span>
+  </button>
+</div>
+<div class="loader" *ngIf="isLoading">
+  <span><i class="fa fa-spinner fa-spin"> </i>Loading...</span>
```
Comment (drizzentic) [inline src/app/hiv-care-lib/surge-report/surge-report-patientlist.component.html]:
```suggestion
  <span><i class="fa fa-spinner fa-spin"></i>Loading...</span>
```

File: src/app/hiv-care-lib/surge-report/surge-report-patientlist.component.ts [hunk 21]
```diff
@@ -0,0 +1,166 @@
+import { Component, OnInit, Input } from '@angular/core';
```
Comment (drizzentic) [inline src/app/hiv-care-lib/surge-report/surge-report-patientlist.component.ts]: Change name of this file to reflect the suggested changes and all other relevant references.
`surge-report-patient-list.component.ts`

File: src/app/hiv-care-lib/surge-report/surge-report-tabular.component.ts [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+        this.indicatorSelected.emit(selectedIndicator);
+      };
+    }
+  }
+  public downloadPdf() {
+
+    const data = document.getElementById('contentToConvert');
+    html2canvas(data).then(canvas => {
+      // Few necessary setting options
+      console.log(canvas);
```
Comment (drizzentic) [inline src/app/hiv-care-lib/surge-report/surge-report-tabular.component.ts]: Remove debug statements

File: src/app/reporting-utilities/report-view/report-view.component.html [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+          </div>
+        </div>
+      </div>
+    </div>
+
+    <div id="contentToConvert" class="panel-body" style="background-color: #ffffff; overflow-x:auto">
+      <table>
+        <tbody *ngFor="let grid of gridOptions.columnDefs">
+          <tr>
+            <!-- <td [attr.colspan]="gridOptions.rowData.length + 1" class="header" ><b>{{grid.headerName}}</b></td> -->
```
Comment (drizzentic) [inline src/app/reporting-utilities/report-view/report-view.component.html]: Remove commented code
```suggestion
            
```

File: src/app/reporting-utilities/report-view/report-view.component.ts [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+  @Output()
+  public CellSelection = new EventEmitter();
+
+  constructor(private domSanitizer: DomSanitizer) { }
+
+  ngOnInit() {
+  }
+  public ngOnChanges(changes: SimpleChanges) {
+    if (changes.SummaryData) {
+      console.log(changes.SummaryData);
```
Comment (drizzentic) [inline src/app/reporting-utilities/report-view/report-view.component.ts]: Remove debug statement
```suggestion
```

File: src/app/reporting-utilities/report-view/report-view.component.ts [hunk 25]
```diff
@@ -0,0 +1,165 @@
...
+      this.sectionIndicatorsValues = this.SummaryData;
+      this.setColumns(this.sectionDefs);
+    }
+  }
+  public setColumns(sectionsData: Array<any>) {
+    console.log(sectionsData);
```
Comment (drizzentic) [inline src/app/reporting-utilities/report-view/report-view.component.ts]:
```suggestion

```

File: src/app/reporting-utilities/report-view/report-view.component.ts [hunk 26]
```diff
@@ -0,0 +1,165 @@
...
+        created.children.push(child);
+      }
+      defs.push(created);
+    }
+    console.log(defs);
+    this.gridOptions.columnDefs = defs;
+  }
+  public setCellSelection(col, val) {
+    const selectedIndicator = { headerName: col.headerName, field: col.field, location: val.location };
+    console.log(selectedIndicator);
```
Comment (drizzentic) [inline src/app/reporting-utilities/report-view/report-view.component.ts]:
```suggestion

```

File: src/app/reporting-utilities/report-view/report-view.component.ts [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+    } else {
+      this.setColumns(this.sectionDefs);
+    }
+  }
+  public downloadPdf() {
+
+    const data = document.getElementById('contentToConvert');
+    html2canvas(data).then(canvas => {
+      // Few necessary setting options
+      console.log(canvas);
```
Comment (drizzentic) [inline src/app/reporting-utilities/report-view/report-view.component.ts]:
```suggestion

```

**Review by drizzentic — APPROVED** [review]

**Review by fatmali — APPROVED** [review]

**Pull request merged**